core/vm/access_list_tracer.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2021 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type accessList struct {
+	addresses map[common.Address]int
+	slots     []map[common.Hash]struct{}
+}
+
+// newAccessList creates a new accessList.
+func newAccessList() *accessList {
+	return &accessList{
+		addresses: make(map[common.Address]int),
+	}
+}
+
+func (al *accessList) AddAddress(address common.Address) {
+	// Set address if not previously present
+	if _, present := al.addresses[address]; !present {
+		al.addresses[address] = -1
+	}
+}
+
+func (al *accessList) AddSlot(address common.Address, slot common.Hash) {
+	idx, addrPresent := al.addresses[address]
+	if !addrPresent || idx == -1 {
+		// Address not present, or addr present but no slots there
+		al.addresses[address] = len(al.slots)
+		slotmap := map[common.Hash]struct{}{slot: {}}
+		al.slots = append(al.slots, slotmap)
+		return
+	}
+	// There is already an (address,slot) mapping
+	slotmap := al.slots[idx]
+	if _, ok := slotmap[slot]; !ok {
+		slotmap[slot] = struct{}{}
+	}
+	// No changes require
+}
+
+func (al *accessList) DeleteAddressIfNoSlotSet(address common.Address) {
+	idx, addrPresent := al.addresses[address]
+	if !addrPresent || idx == -1 {
+		delete(al.addresses, address)
+	}
+}
+
+// Copy creates an independent copy of an accessList.
+func (a *accessList) Copy() *accessList {
+	cp := newAccessList()
+	for k, v := range a.addresses {
+		cp.addresses[k] = v
+	}
+	cp.slots = make([]map[common.Hash]struct{}, len(a.slots))
+	for i, slotMap := range a.slots {
+		newSlotmap := make(map[common.Hash]struct{}, len(slotMap))
+		for k := range slotMap {
+			newSlotmap[k] = struct{}{}
+		}
+		cp.slots[i] = newSlotmap
+	}
+	return cp
+}
+
+func (a *accessList) ToAccessList() *types.AccessList {
+	acl := make([]types.AccessTuple, 0, len(a.addresses))
```

Comment:
```suggestion
	acl := make(types.AccessList, 0, len(a.addresses))
```
And then no cast is needed, just return it

Reply:
Another idea would be to make `state.accessList`  public. Although I do think it's kind of nice to have the tracer be standalone, even if it does introduce this copy-paste struct. @fjl what do you think?

Reply:
Yeah I thought about that too, but I think having two public structs called `AccessList` would probably confuse a lot of people, especially since it is only used as an internal datastructure